File: Python/rt/projectile.py
from rt.tuple import *
from rt.color import *
from rt.canvas import *

class Projectile(object):
    def __init__(self, pos, vel):
        self.Position = pos
        self.Velocity = vel

class Environment(object):
    def __init__(self, g, w):
        self.Gravity = g
        self.Wind = w

def Tick(env, proj):
    p = proj.Position + proj.Velocity
    v = proj.Velocity + env.Gravity + env.Wind

    return Projectile(Point(p.x, p.y, p.z), Vector(v.x, v.y, v.z))

if __name__ == '__main__':
    print('Fire the cannon!')

    start = Point(0, 1, 0)
    velocity = Vector(1, 1, 0).normalize * 10
    p = Projectile(start, velocity)

    gravity = Vector(0, -0.1, 0)
    wind = Vector(-0.01, 0, 0)
    e = Environment(gravity, wind)

    c = Canvas(900, 550)
    color = Color(1, 0, 0)
    ticks = 0

    while (p.Position.y >= 0):
        print(f'Tick {ticks}:  {p.Position}')

        y = c.Height - int(round(p.Position.y))
        x = int(round(p.Position.x))
        c[[x, y]] = color

        ticks += 1
        p = Tick(e, p)

    file = open(r'projectile.ppm', 'w')
    file.write(c.GetPPM())
    file.close()

    print('Done!')

File: C/src/projectile.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "tuple.h"
#include "color.h"
#include "canvas.h"

/***********************************************
 * A project that uses the ray tracer challenge
 * library to simulate a projectile.
 ***********************************************/

typedef struct {
    Tuple Position;
    Tuple Velocity;
} Projectile;

typedef struct
{
    Tuple Gravity;
    Tuple Wind;
} Environment;

Projectile Tick(Environment env, Projectile proj)
{
    Tuple position = addt(proj.Position, proj.Velocity);
    Tuple velocity = addt(proj.Velocity, addt(env.Gravity, env.Wind));

    Projectile a = {
                        { position.x, position.y, position.z },
                        { velocity.x, velocity.y, velocity.z }
                    };
    return a;
}

int main (int argc, char *argv[]) {

    printf("Fire the cannon!\r\n");
 
    Tuple v = {1, 1, 0};
    Tuple start = {0, 1, 0};
    Tuple velocity = mult(normv(v), 10);
    Projectile p = { start, velocity };

    Tuple gravity = {0, -0.1, 0};
    Tuple wind = {-0.01, 0, 0 };
    Environment e = { gravity, wind };
    
    Canvas c = create_canvas(900, 550);
    Color color = create_color(1, 0, 0);
    int ticks = 0;

    do
    {
        char *tuple = tuple_string(p.Position);
        printf("Tick %i: %s\r\n", ticks++, tuple);
        free(tuple);
        
        int y = c.height - round(p.Position.y);
        int x = round(p.Position.x);
        setPixel(c, x, y, color);

        p = Tick(e, p);
    } while (p.Position.y >= 0);

    FILE *f = fopen("projectile.ppm", "w");
    fprintf(f, getPPM(c));
    fclose(f);

    printf("Done!\r\n\r\n");

    return 0;
}